src/proxy_protocol/mod.rs

```diff
@@ -0,0 +1,571 @@
+    let header = HeaderResult::parse(&buffer[..full_length]);
+
+    match header {
+        HeaderResult::V1(Ok(_header)) => Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "V1 Proxy Protocol header detected when parsing data",
+        )),
+        HeaderResult::V2(Ok(header)) => {
+            let client_address = match header.addresses {
+                v2::Addresses::IPv4(ip) => {
+                    SocketAddr::new(IpAddr::V4(ip.source_address), ip.source_port)
+                }
+                v2::Addresses::IPv6(ip) => {
+                    SocketAddr::new(IpAddr::V6(ip.source_address), ip.source_port)
+                }
+                v2::Addresses::Unix(unix) => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::InvalidData,
+                        format!(
+                            "Unix socket addresses are not supported. Addresses: {:?}",
+                            unix
+                        ),
+                    ));
+                }
+                v2::Addresses::Unspecified => {
+                    // V2 PROXY header addresses "Unspecified"
+                    // Return client address as `None` so that "unknown" is used in the http header
+                    return Ok((stream, None));
+                }
+            };
+
+            Ok((stream, Some(client_address)))
+        }
+        HeaderResult::V1(Err(_error)) => Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "No valid V1 Proxy Protocol header received",
+        )),
+        HeaderResult::V2(Err(_error)) => Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "No valid V2 Proxy Protocol header received",
+        )),
+    }
+}
+
+/// Middleware for adding client IP address to the request `forwarded` header.
+/// see spec: <https://www.rfc-editor.org/rfc/rfc7239#section-5.2>
+#[derive(Debug)]
+pub struct ForwardClientIp<S> {
+    inner: S,
+    client_address_opt: Option<SocketAddr>,
+}
+
+impl<S> Service<Request<hyper::Body>> for ForwardClientIp<S>
+where
+    S: Service<Request<hyper::Body>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<hyper::Body>) -> Self::Future {
+        let forwarded_string = match self.client_address_opt {
+            Some(socket_addr) => match socket_addr {
+                SocketAddr::V4(addr) => {
+                    format!("for={}:{}", addr.ip(), addr.port())
+                }
+                SocketAddr::V6(addr) => {
+                    format!("for=\"[{}]:{}\"", addr.ip(), addr.port())
+                }
+            },
+            None => "for=unknown".to_string(),
+        };
+
+        if let Ok(header_value) = HeaderValue::from_str(&forwarded_string) {
```

Comment:
If the header is already set, we should also consider what to do. Probably overwrite, but maybe something to consider in that header spec.